src/common/guac_surface.c

```diff
@@ -77,6 +79,446 @@
+            if (intersects == 2) {
+
+                /* Exclude this from the normal refresh as it is completely
+                 * covered by the lossy dirty rectangle. */
+                current->flushed = 1;
+
+            }
+
+            /* Partial intersection.
+             * The rectangle will be split if there is room on the queue. */
+            else if (intersects == 1 &&
+                     surface->bitmap_queue_length < GUAC_COMMON_SURFACE_QUEUE_SIZE-5) {
+
+                /* Clip and split rectangle into rectangles that are outside the
+                 * lossy rectangle which are added to the normal refresh queue.
+                 * The remaining rectangle which overlaps with the lossy
+                 * rectangle is marked flushed to not be refreshed in the normal
+                 * refresh cycle.
+                 */
+                guac_common_rect split_rect;
+                while (guac_common_rect_clip_and_split(&current->rect,
+                       &surface->lossy_dirty_rect, &split_rect)) {
+
+                    /* Add new rectangle to update queue */
+                    __guac_common_surface_flush_rect_to_queue(surface,
+                                                              &split_rect);
+
+                }
+
+                /* Exclude the remaining part of the dirty rectangle
+                 * which is completely covered by the lossy dirty rectangle. */
+                current->flushed = 1;
+
+            }
+            current++;
+
+        }
+
+        /* Flush the lossy bitmap */
+        __guac_common_surface_flush_to_bitmap_impl(surface,
+                                                 &surface->lossy_dirty_rect, 1);
+
+        /* Flag this area as lossy so it can be moved back to the
+         * dirty rect and refreshed normally when refreshed less frequently. */
+        int x = surface->lossy_dirty_rect.x;
+        int y = surface->lossy_dirty_rect.y;
+        int w = (x + surface->lossy_dirty_rect.width) / GUAC_COMMON_SURFACE_STAT_RECT;
+        int h = (y + surface->lossy_dirty_rect.height) / GUAC_COMMON_SURFACE_STAT_RECT;
+        x /= GUAC_COMMON_SURFACE_STAT_RECT;
+        y /= GUAC_COMMON_SURFACE_STAT_RECT;
+
+        for (int j = y; j <= h; j++) {
+            for (int i = x; i <= w; i++) {
+                surface->lossy_rect[j][i] = 1;
+            }
+        }
+
+        /* Clear the lossy dirty flag. */
+        surface->lossy_dirty = 0;
+    }
+
+}
+
+/**
+ * Calculate the current average update frequency for a given area on the
+ * surface.
+ *
+ * @param surface
+ *     The surface on which the update frequency will be calculated.
+ *
+ * @param x
+ *     The x coordinate for the area.
+ *
+ * @param y
+ *     The y coordinate for the area.
+ *
+ * @param w
+ *     The area width.
+ *
+ * @param h
+ *     The area height.
+ *
+ * @return
+ *     The average update frequency.
+ */
+static double __guac_common_surface_calculate_update_freq(
```

Comment:
Why `double`? Is there a way to do this without using floating point arithmetic?